### DISCUSSION TITLE
TypeError: Cannot read properties of undefined (reading 'timeoutType')

### DISCUSSION THREAD

**Discussion by magnusrodseth** [trpc/trpc#4112] [Q&A]
To set the stage, I am using a Turborepo with Next.js, tRPC and Clerk for authentication and user management.

Ever since the start of my project, I sometimes get this error message:

```sh
@acme/web:dev: error - unhandledRejection: TypeError: Cannot read properties of undefined (reading 'timeoutType')
@acme/web:dev:     at _resume (evalmachine.<anonymous>:11000:33)
@acme/web:dev:     at resume (evalmachine.<anonymous>:10969:7)
@acme/web:dev:     at connect (evalmachine.<anonymous>:10956:7)
@acme/web:dev: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
@acme/web:dev: TypeError: Cannot read properties of undefined (reading 'timeoutType')
@acme/web:dev:     at _resume (evalmachine.<anonymous>:11000:33)
@acme/web:dev:     at resume (evalmachine.<anonymous>:10969:7)
@acme/web:dev:     at connect (evalmachine.<anonymous>:10956:7)
```

The error happens sporadically, and I have not been able to pinpoint any pattern of it occurring. The thing that baffles me is that when I Google this error, no other discussions around it exist.

It may happen because I throw `TRPCError` in my server code, but shouldn't that just be handled using the `error` object in `useQuery` / `useMutation` on the client? There should be no need to manually use `.catch()`.

The stack trace is also very challenging to navigate, and does not offer much help. Is `timeoutType` something tRPC-specific, or is it just related to `node.js`?

Do you have any general recommendations on how to debug and fix this issue?

**Comment by kamilogorek** [comment]
> The stack trace is also very challenging to navigate, and does not offer much help. Is timeoutType something tRPC-specific, or is it just related to node.js?

It's Node.js, and to be more specific, it's Unidici, their HTTP client. There are only 2 places where it can break there - https://github.com/nodejs/undici/blob/dc8d11183fb95aa93d296bde3e21501856849367/lib/client.js#L1207-L1220

I'd ask in their repo to see if they have an idea under what circumstances this can be triggered.

**Reply by magnusrodseth** [reply]
@kamilogorek Do you know if tRPC uses Undici? It seems like my sometimes crashes when I send many requests in a short interval, for instance by refreshing my page many times over.

I have not manually installed Undici, so it must be a peer dependency.

**Reply by kamilogorek** [reply]
Undici _is_ Node's fetch API :) so yes, tRPC is using it, as it's using built-in `fetch` API.

**Reply by magnusrodseth** [reply]
@kamilogorek Do you know if tRPC has an adapter for the fetch API, enabling me to switch to `axios` for instance? I wasn't able to find any docs on it.

**Reply by kamilogorek** [reply]
I assume this is a client part of the setup? Then: 

```ts
const client = createTRPCProxyClient<AppRouter>({
  links: [
    httpLink({
      url: 'http://localhost:3000',
      fetch: yourFetchImpl
    }),
  ],
});
```

The same goes for `httpBatchLink`. Server doesn't perform any `fetch` calls by itself.
